---
title: "Intro_to_MAF"
date: "2024-01-19"
---

***Deliverables***
-Upload this R Notebook to your GitHub and submit the link to your Repo on BB.
-Include any graphs or figures created in this assignment in the folder with your R notebook with descriptive file names.

Since this is an optional partner activity, it is okay if your answers are the same as your partner’s as long as everyone understands it and could explain it in their own words if asked. Each person must individually push their code to Github. *At the top of your R Notebook, write the name of you and your partner(s) as a comment.*

***Complete the following coding activity and answer any following questions as comments in your R Notebook***

In MAF_Tutorial.Rmd, you looked at how to analyze mutation data in terms of clinical variables as well as mutation status. In this assignment, you will need to combine the two skills to demonstrate your understanding of categorical variables and R data structures.

*Pre-Assignment*
Load all necessary packages, read in the clinical data .csv file you have in your analysis_data folder, and instantiate the MAF_object.
```{r}
knitr::opts_knit$set(root.dir = normalizePath("/Users/amansharma/Desktop/QBIO490/qbio_490_amansharma/analysis_data"))

library(BiocManager)
library(TCGAbiolinks)
library(maftools)

clinical <- read.csv("/Users/amansharma/desktop/qbio490/qbio_490_amansharma/analysis_data/skcm_clinical_data.csv")
colnames(clinical)[ colnames(clinical) == "bcr_patient_barcode" ] <- "Tumor_Sample_Barcode"
write.csv(clinical, "/Users/amansharma/Desktop/QBIO490/qbio_490_amansharma/analysis_data/skcm_clinical_data.csv", row.names = FALSE)

```

```{r}
maf_query <- GDCquery(
    project = "TCGA-SKCM", 
    data.category = "Simple Nucleotide Variation", 
    access = "open", # we only have access to somatic mutations which are open access
    data.type = "Masked Somatic Mutation", 
    workflow.type = "Aliquot Ensemble Somatic Variant Merging and Masking"
)

#GDCdownload(maf_query) # you shouldn't have to redownload any data!

maf <- GDCprepare(maf_query) # as long as it runs, ignore any errors

maf_object <- read.maf(maf = maf, 
                       clinicalData = clinical,
                       isTCGA = TRUE)


```


*1*
Choose a clinical variable to separate your populations into two different groups and rewrite the column or create a new column with that variable as a factor. ***Do not use age or vital_status as your clinical variable.*** Hint: if your variable is continuous, you will need to determine your own cutoffs for the different levels of the factor. If your variable is categorical and has more than two possible values, choose the two that are the most common.
```{r}
# I counted ICD_10 codes

# Based on this, I think we can separate based on ICD codes C77.3 and C44.5 as they are most common.

maf_object@clinical.data$icd_group <- ifelse(maf_object@clinical.data$icd_10 %in% c("C77.3", "C44.5"), maf_object@clinical.data$icd_10, "Other")

# Generate an oncoplot that includes the ICD-10 group as a clinical feature


```


*2*
Create a co-oncoplot with the top 10-20 (you choose) most mutated genes for the two groups. Pick one that has a large discrepancy in % mutated or type of mutations between the groups and research it. 
Research it. What is the gene used for? Can you think of any reason for the discrepancy?
```{r}

C773_patient_barcodes <- maf_object@clinical.data[maf_object@clinical.data$icd_group == 'C77.3',]$Tumor_Sample_Barcode
C773_maf <- subsetMaf(maf = maf_object, tsb = C773_patient_barcodes)

# Subsetting MAF for patients with ICD-10 code C44.5
C445_patient_barcodes <- maf_object@clinical.data[maf_object@clinical.data$icd_group == 'C44.5',]$Tumor_Sample_Barcode
C445_maf <- subsetMaf(maf = maf_object, tsb = C445_patient_barcodes)

# Create co-oncoplot comparing the two ICD-10 groups
coOncoplot(m1 = C773_maf, 
           m2 = C445_maf, 
           m1Name = "ICD-10: C77.3", 
           m2Name = "ICD-10: C44.5", 
           borderCol = NA)

# The one that had the largest discrepancy is DNAH5 with an overall discrepancy of 19%.

#DNAH5 is a protein coding gene that is essential for structure and function of cilia, which are essential for motility, fluid movement along cell surface, and signaling. They have been considered "arms" of axonemal dyenins, which provide the force needed for cilia to function. The gene is also involved in PCD, which involves chronic respiratory tract infections, abnormal organ positioning, and fertility issues due to defective cilia.

#Thus, mutations in DNAH5 are tied to PCD development due to how they affect ciliary motion of cells. Mutations may occur due to environmental damage or signaling issues, contributing to these discrepancies.

```

*3*
Create a contingency table with your variable and chosen gene. Run a Fisher’s Exact Test between presence of mutations for that gene and your clinical variable. Create and save a mosaic plot. 
Interpret the output of the Fisher’s Exact Test in terms of the odds ratio and p-value.
```{r}

# Load necessary library
library(maftools)

dna5_mutations <- subsetMaf(maf = maf_object, genes = 'DNAH5', mafObj = TRUE)
maf_object@clinical.data$DNAH5_mutation <- maf_object@clinical.data$Tumor_Sample_Barcode %in% dna5_mutations@data$Tumor_Sample_Barcode
contingency_table <- table(maf_object@clinical.data$icd_group, maf_object@clinical.data$DNAH5_mutation)
print(contingency_table)
fishers_test <- fisher.test(contingency_table)
print(fishers_test)
mosaicplot(contingency_table, main = "Mosaic Plot of DNAH5 Mutations vs ICD-10 Code", color = TRUE)
print(paste("Odds Ratio: ", fishers_test$estimate))
print(paste("P-value: ", fishers_test$p.value))



```


*4*
Subset your maf_object based on your chosen clinical variable and create a co-lollipop plot of your chosen gene divided between the two different clinical variable possibilities. Include descriptive names on your plot.
Do you notice any difference in terms of mutations (e.g. sites, types, number) between the two populations?
```{r}

library(maftools)

# Extract clinical data from MAF object
maf_clinical <- maf_object@clinical.data
na_mask <- ifelse(maf_clinical$icd_10 == '[Not Available]', FALSE, TRUE)
maf_clinical <- maf_clinical[na_mask,]
maf_clinical$icd_category <- ifelse(maf_clinical$icd_10 == 'C77.3', 'C77.3', ifelse(maf_clinical$icd_10 == 'C44.5', 'C44.5', 'Other'))
C773_barcodes <- maf_clinical[maf_clinical$icd_category == 'C77.3',]$Tumor_Sample_Barcode
C445_barcodes <- maf_clinical[maf_clinical$icd_category == 'C44.5',]$Tumor_Sample_Barcode
maf_C773 <- subsetMaf(maf = maf_object, tsb = C773_barcodes)
maf_C445 <- subsetMaf(maf = maf_object, tsb = C445_barcodes)

lollipopPlot2(m1 = maf_C445, 
              m2 = maf_C773, 
              m1_name = "C44.5",
              m2_name = "C77.3",
              gene = "DNAH5")

#C77.3 seems to have a lot more missense mutations on variety of sectors of the gene than C44.5, on basically every sector too. There is also a noticeable, but small, increase in nonsense mutations for patients with icd_10 code C77.3

```


*5*
Create your Overall_Survival_Status column and create a mafSurvival KM plot based on mutations in your chosen gene.
Does there seem to be a difference? Hypothesize why or not based on the other analysis you did with the gene above.
```{r}

library(maftools)
maf_object@clinical.data$Overall_Survival_Status <- ifelse(maf_object@clinical.data$vital_status == "Dead", 1, 0)  # 1 for deceased, 0 for alive
maf_object@clinical.data$DNAH5_mut_status <- ifelse(maf_object@clinical.data$Tumor_Sample_Barcode %in% subsetMaf(maf = maf_object, genes = 'DNAH5', mafObj = TRUE)@data$Tumor_Sample_Barcode, "Mutated", "Wildtype")
mafSurvival(maf = maf_object, 
            genes = 'DNAH5', 
            time = 'death_days_to', 
            Status = 'Overall_Survival_Status', 
            isTCGA = TRUE)

# There does seem to be a very slight difference. 
```